src/sink/gcs.rs

```diff
@@ -0,0 +1,345 @@
+use simd_json::json;
+use tremor_pipeline::{EventId, OpMeta};
+use tremor_value::Value;
+
+pub struct GoogleCloudStorage {
+    config: Config,
+    remote: Option<Client>,
+    is_down: bool,
+    qos_facility: Box<dyn SinkQoS>,
+    reply_channel: Option<Sender<sink::Reply>>,
+    is_linked: bool,
+}
+
+#[derive(Deserialize)]
+pub struct Config {
+    pem: String,
+}
+
+enum StorageCommand {
+    Create(String, String),
+    Add(String, String, String),
+    Remove(String, String),
+    ListBuckets(String),
+    Fetch(String, String),
+    Download(String, String),
+    Delete(String),
+    ListObjects(String),
+    Unknown,
+}
+
+impl std::fmt::Display for StorageCommand {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match *self {
+                StorageCommand::Create(_, _) => "create_bucket",
+                StorageCommand::Add(_, _, _) => "add_object",
+                StorageCommand::Remove(_, _) => "remove_object",
+                StorageCommand::ListBuckets(_) => "list_buckets",
+                StorageCommand::Fetch(_, _) => "get_object",
+                StorageCommand::Download(_, _) => "download_object",
+                StorageCommand::Delete(_) => "delete_bucket",
+                StorageCommand::ListObjects(_) => "list_objects",
+                StorageCommand::Unknown => "Unknown",
+            }
+        )
+    }
+}
+
+impl ConfigImpl for Config {}
+
+impl offramp::Impl for GoogleCloudStorage {
+    fn from_config(config: &Option<OpConfig>) -> Result<Box<dyn Offramp>> {
+        if let Some(config) = config {
+            let config: Config = Config::new(config)?;
+            let headers = HeaderMap::new();
+            let remote = Some(block_on(auth::json_api_client(&config.pem, &headers))?);
+            let hostport = "storage.googleapis.com:443";
+            Ok(SinkManager::new_box(Self {
+                config,
+                remote,
+                is_down: false,
+                qos_facility: Box::new(QoSFacilities::recoverable(hostport.to_string())),
+                reply_channel: None,
+                is_linked: false,
+            }))
+        } else {
+            Err("Offramp Google Cloud Storage requires a config".into())
+        }
+    }
+}
+
+fn parse_command(value: &Value<'_>) -> Result<StorageCommand> {
+    if let Value::Object(o) = value {
+        let cmd_name: &str = if let Some(Value::String(cmd_name)) = o.get("command") {
+            cmd_name
+        } else {
+            return Err("Invalid Command, expected `command` field".into());
+        };
+
+        let command = match cmd_name {
+            "fetch" => StorageCommand::Fetch(
+                if let Some(Value::String(bucket_name)) = o.get("bucket") {
+                    bucket_name.to_string()
+                } else {
+                    return Err("Invalid Command, expected `bucket` field".into());
+                },
```

Comment:
Since this pattern repeats quite a lot here a suggestion how it can be re-written to make it a bit easier to read, especially since it's in a struct field:

```suggestion
                o.get_str("bucket").ok_or("Invalid Command, expected `bucket` field")?.to_string()
```

Reply:
Yep - we'll hit these in the next revision as live testing in a real environment will no doubt uncover some enhancements before we mark this as stable ...